### DISCUSSION TITLE
Comparison to waPC? (and other WASM RPC tools?)

### DISCUSSION THREAD

**Discussion by agilgur5** [bytecodealliance/wrpc#495] [Q&A] [answered]
Hi there, I discovered wRPC recently but have previously used [waPC](https://github.com/wapc), so was wondering if someone could compare/contrast the two? And possibly other such WASM RPC tools

Afaict, wRPC is grounded in the component specification and WIT, so using newer standards.
The host/guest implementations seem vaguely similar from a quick glance, although wRPC focuses on generating WIT bindings rather than needing a guest SDK?
Very back of the napkin analysis, please correct me! 🙏 

I'm primarily looking at this in the lens of compatibility, composition, and cross-platform usability between various WASM tools and have been contributing to the ecosystem a bit to improve that

**EDIT**: Also linked this in [`#kubewarden` on k8s Slack](https://kubernetes.slack.com/archives/C01T3GTC3L7/p1732212581797809)

**Comment by rvolosatovs** [comment, marked as answer]
I have just learned about waPC and took a very brief look at the [spec](https://wapc.io/docs/spec) - from what I currently understand, waPC project only focuses on Wasm *module* <-> host communication. Note, that as written in the spec, the protocol is fundamentally incompatible with [component model](https://component-model.bytecodealliance.org/) for Wasm component <-> host communication, due to the lack of shared memory access, at least today. It seems that waPC is, in fact, competing with [component model](https://component-model.bytecodealliance.org/), since there appears to be no need for waPC for Wasm components

wRPC is a general purpose RPC framework, Wasm *component* <-> host communication is out-of-scope (but possible) for wRPC, on the contrary, wRPC is mostly useful for the *hosts* themselves, but Wasm *components* could use it as well. A typical use case would be a Wasm component using [component model](https://component-model.bytecodealliance.org/) to call a host export and the host itself calling some other party using wRPC to provide an implementation (e.g. a plugin). wRPC, built on WIT, provides a direct mapping of the component call to the remote call allowing for optimized and ergonomic data transfer.

TL;DR
waPC is for Wasm module <-> Wasm host communication in-memory
wRPC is for anything <-> anything communication over a variety of different transports: in-memory, TCP, QUIC, HTTP3, NATS.io etc.

**Reply by agilgur5** [reply]
Thanks for the detailed answer!

### waPC 

> waPC project only focuses on Wasm _module_ <-> host communication

Yes, that's correct as far as I understand.

> It seems that waPC is, in fact, competing with [component model](https://component-model.bytecodealliance.org/), since there appears to be no need for waPC for Wasm components

Yes, waPC pre-dates the component model specification and [may very well become deprecated](https://kubernetes.slack.com/archives/C01T3GTC3L7/p1732808534124799?thread_ts=1732212581.797809&cid=C01T3GTC3L7) in favor of the component model once it matures. It might even be mature enough with WASI 0.2.0 etc these days; bindings generation in various languages might be the only piece that could use more tooling if my understanding is correct.

### wRPC

> Wasm _component_ <-> host communication is out-of-scope (but possible) for wRPC, on the contrary, wRPC is mostly useful for the _hosts_ themselves, but Wasm _components_ could use it as well.

Ahh I see. The scopes did seem different from an initial read, but I couldn't quite tell. Thanks for clarifying this!

> Wasm component using [component model](https://component-model.bytecodealliance.org/) to call a host export and the host itself calling some other party using wRPC to provide an implementation (e.g. a plugin)

This is a very helpful example use-case!

These could be potentially useful to add to the docs in some way for other folks like me looking at WASM RPC tools and trying to understand how they all fit in together. I definitely did not grok this from my reading of the README 😅
